### DISCUSSION TITLE
Prisma + Remix + Architect - impossible to make it work?

### DISCUSSION THREAD

**Discussion by daveslutzkin** [prisma/prisma#22527] [Q&A] [answered]
### Question

I'm having a lot of trouble trying to make Prisma work with a fresh Remix 2 project based on the Architect template.

Basically the problem seems to be that Architect "bundles" the server in a subdirectory of the project root, which means that prisma's postinstall hook never gets a chance to run on the node_modules which will be deployed.

### How to reproduce (optional)

See the repo here: https://github.com/daveslutzkin/remix-arc-prisma.

This is a complete toy app with a SQLite db and a single table in the schema.

`.env`
```
DATABASE_URL="file:test.db"
```

Try to run the local dev server using `npm run dev` and it should work on [`http://localhost:3333`](http://localhost:3333/).

Then try to deploy:

- `npm run build`
- `npx arc deploy`

As soon as architect has finished hydrating the server, if you refresh the [`http://localhost:3333`](http://localhost:3333/) page you'll see that it starts failing with this error:

```
@prisma/client did not initialize yet. Please run "prisma generate" and try to import it again. In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
```

(And of course the Lambda when it tries to serve a request gets the same error.)

This is because architect creates a separate copy of `node_modules` in the `server/` directory, so that it only bundles/deploys what the server specifically needs.

But when the `prisma` `postinstall` script runs it generates the client into `project_root/node_modules/.prisma/client` rather than `project_root/server/node_modules/.prisma/client`. So the version of the code that's deployed doesn't have a prisma client. (And once the server/node_modules directory exists, that's also what the local dev server uses.)

### Expected behavior (optional)

I hesitate to call this a problem only with prisma, though this issue doesn't seem to arise for any other package. It's feels like it's at least partly architect's fault, but that said surely it's ok to have two copies of `node_modules` in a tree?

Anyway, the expected behaviour I suppose is that when the postinstall hook runs on the prisma package installed into `project_root/server/node_modules/@prisma`, it generates a client into a path relative to that same package, so `project_root/server/node_modules/.prisma/client`.

The obvious solution here is to use a custom location for the client, but then we bump into issues with ESM which I don't fully understand - will follow up with a comment describing that.

**Comment by daveslutzkin** [comment]
So the next step is of course to try to use a custom location for the client. In that case I'd include the client in version control and it would be bundled separately as part of the server bundle, which makes sense to me. (I don't mind having the client in version control personally.)

Pick a logical location for the client into `prisma/schema.prisma`:
```
generator client {
  provider = "prisma-client-js"
  output   = "./generated/client"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model User {
  id    Int     @id @default(autoincrement())
}
```

Then change `app/db.server.ts` to point to that location (as the generator advises me):
```
import { PrismaClient } from "./prisma/generated/client"
const db = new PrismaClient()
export { db }
```

Dev server says `Could not resolve "./prisma/generated/client"` because it's running the server from `server/` so it's a relative path, change it to `../prisma/generated/client`.

Then we get this error:
```
Dynamic require of "tty" is not supported
```

So now we've bumped into https://github.com/prisma/prisma/issues/5030, which is that Prisma doesn't support ESM modules out of the box.

I don't know enough about JS imports/modules/whatever to make this work.

Is there a simple solution to actually make prisma work with a pretty simple combo of remix and architect?

**Comment by daveslutzkin** [comment, marked as answer]
I ended up hacking this by patching architect (using `patch-package`) to add a new hook:
```
diff --git a/node_modules/@architect/deploy/src/sam/index.js b/node_modules/@architect/deploy/src/sam/index.js
index 626ea56..56cd8fc 100644
--- a/node_modules/@architect/deploy/src/sam/index.js
+++ b/node_modules/@architect/deploy/src/sam/index.js
@@ -140,6 +140,11 @@ module.exports = function samDeploy (params, callback) {
       else callback()
     },

+    // deploy.postHydrate plugins
+    function runPostHydratePlugins (callback) {
+      plugins.postHydrate({ inventory }, callback)
+    },
+
     // Print a size report
     function chonkyBois (callback) {
       sizeReport({ inventory, update }, callback)
diff --git a/node_modules/@architect/deploy/src/sam/plugins/index.js b/node_modules/@architect/deploy/src/sam/plugins/index.js
index de536cc..58dfa8b 100644
--- a/node_modules/@architect/deploy/src/sam/plugins/index.js
+++ b/node_modules/@architect/deploy/src/sam/plugins/index.js
@@ -1,6 +1,7 @@
 let start = require('./start')
 let services = require('./services')
+let postHydrate = require('./postHydrate')
 let target = require('./target')
 let end = require('./end')

-module.exports = { start, services, target, end }
+module.exports = { start, services, postHydrate, target, end }
diff --git a/node_modules/@architect/deploy/src/sam/plugins/postHydrate.js b/node_modules/@architect/deploy/src/sam/plugins/postHydrate.js
new file mode 100644
index 0000000..899ef50
--- /dev/null
+++ b/node_modules/@architect/deploy/src/sam/plugins/postHydrate.js
@@ -0,0 +1,30 @@
+let { deepFrozenCopy } = require('@architect/utils')
+
+/**
+ * deploy.postHydrate plugins
+ */
+module.exports = function postHydratePlugins (params, callback) {
+  let deployPostHydratePlugins = params.inventory.inv.plugins?._methods?.deploy?.postHydrate
+  if (deployPostHydratePlugins) {
+    let inventory = deepFrozenCopy(params.inventory)
+    let { arc } = inventory.inv._project
+    async function runPlugins () {
+      for (let plugin of deployPostHydratePlugins) {
+        let { _type } = plugin
+        let result
+        // Plugins accept an option object
+        if (_type === 'plugin') {
+          result = await plugin({ arc, inventory })
+        }
+        // Legacy macros use ordered args
+        if (_type === 'macro') {
+          result = await plugin(arc, inventory)
+        }
+      }
+    }
+    runPlugins()
+      .then(() => callback(null))
+      .catch(callback)
+  }
+  else callback(null)
+}
```

This new hook then does this:
```
  deploy: {
    async postHydrate() {
      spawnSync("rm", ["-rf", "server/node_modules/.prisma"]);
      spawnSync("cp", ["-r", "node_modules/.prisma", "server/node_modules/.prisma"]);
    },
  },
```

This is obviously a hack and not ideal but it gets me over the hump, I hope it helps others.

**Comment by shftlvch** [comment]
@daveslutzkin ty for the hack. I was considering just using CDK + Fargate bc of this issue. Have you tried SST/Serverless instead?

**Reply by daveslutzkin** [reply]
Unfortunately I ended up using fly.io because it just worked. I haven't tried SST/Serverless, surely should just work!

**Reply by sarathy-partha** [reply]
@daveslutzkin, I did it this way. See if it helps.

Add Shared module to Arc template as follows,
```
@shared
src "./node_modules/.prisma"
``` 

`npx prisma generate` continues to generate into node_modules/.prisma

`import { PrismaClient } from "@architect/shared/client/index.js";` in db.server.ts

and finally generate the schema as postinstall in package.json
`"postinstall": "npx prisma generate"`
